### PR TITLE
script for storage backend migrations

### PR DESCRIPTION
**Pull request by felixjogris**
This is a follow-up to https://github.com/PrivateBin/PrivateBin/discussions/1007
It basically adds bin/migrate.php which is a command line tool for copying or moving data between storage backends. Source and destination backends may be of different or same type. If they are of same type, their configuration must differ.
For this to work, I had to drop singleton behaviour from all storage backends and their unit tests. This also means all static properties and static member functions from AbstractData and derived classes are now instance members - with the exception of upgradePreV1Format().

### CODE REVIEW

File: lib/Data/Database.php [hunk 1]
```diff
@@ -535,19 +525,36 @@ protected function _getExpiredPastes($batchsize)
...
+        $pastes = array();
+        $rows   = $this->_select(
+            'SELECT "dataid" FROM "' . $this->_sanitizeIdentifier('paste') . '"',
+            array()
+        );
+        if (is_array($rows) && count($rows)) {
+            foreach ($rows as $row) {
+                $pastes[] = $row['dataid'];
+            }
+        }
```
Comment (elrido) [inline lib/Data/Database.php:542]: The `_select()` function does some extra bits for the OCI driver for converting CLOB type columns from streams to strings, but the dataid will be a `CHAR(16)` in all drivers, so we could get away just fetching only the one row, straight:
```suggestion
        $pastes = $this->_db->query(
            'SELECT "dataid" FROM "' . $this->_sanitizeIdentifier('paste') . '"'
        )->fetchAll(PDO::FETCH_COLUMN, 0);
```
... and I should apply the same change to `_getExpiredPastes()`.

**Review by elrido — CHANGES_REQUESTED** [review]
I can do the change in `_getExpiredPastes()` in a follow up pull request, where I also take care of the other discussed items (unit test, moving the other scripts to bin, etc.)

**Review by elrido — CHANGES_REQUESTED** [review]
There are also many unit tests in the ControllerTest and it's siblings are failing, usually after creating the pastes and then checking if they exist.

**Comment by elrido** [comment]
For the Database the solution is to keep the cache static:
```diff
diff --git a/lib/Data/AbstractData.php b/lib/Data/AbstractData.php
index eddab09..b28fd53 100644
--- a/lib/Data/AbstractData.php
+++ b/lib/Data/AbstractData.php
@@ -108,7 +108,7 @@ abstract class AbstractData
         if ($namespace === 'traffic_limiter') {
             foreach ($this->_last_cache as $key => $last_submission) {
                 if ($last_submission <= $time) {
-                    unset($thi->_last_cache[$key]);
+                    unset($this->_last_cache[$key]);
                 }
             }
         }
diff --git a/lib/Data/Database.php b/lib/Data/Database.php
index f431858..361412f 100644
--- a/lib/Data/Database.php
+++ b/lib/Data/Database.php
@@ -30,7 +30,7 @@ class Database extends AbstractData
      *
      * @var array
      */
-    private $_cache = array();
+    private static $_cache = array();
 
     /**
      * instance of database connection
@@ -150,12 +150,12 @@ class Database extends AbstractData
     public function create($pasteid, array $paste)
     {
         if (
-            array_key_exists($pasteid, $this->_cache)
+            array_key_exists($pasteid, self::$_cache)
         ) {
-            if (false !== $this->_cache[$pasteid]) {
+            if (false !== self::$_cache[$pasteid]) {
                 return false;
             } else {
-                unset($this->_cache[$pasteid]);
+                unset(self::$_cache[$pasteid]);
             }
         }
 
@@ -222,11 +222,11 @@ class Database extends AbstractData
      */
     public function read($pasteid)
     {
-        if (array_key_exists($pasteid, $this->_cache)) {
-            return $this->_cache[$pasteid];
+        if (array_key_exists($pasteid, self::$_cache)) {
+            return self::$_cache[$pasteid];
         }
 
-        $this->_cache[$pasteid] = false;
+        self::$_cache[$pasteid] = false;
         try {
             $paste = $this->_select(
                 'SELECT * FROM "' . $this->_sanitizeIdentifier('paste') .
@@ -242,10 +242,10 @@ class Database extends AbstractData
         $data       = Json::decode($paste['data']);
         $isVersion2 = array_key_exists('v', $data) && $data['v'] >= 2;
         if ($isVersion2) {
-            $this->_cache[$pasteid] = $data;
+            self::$_cache[$pasteid] = $data;
             list($createdKey)       = $this->_getVersionedKeys(2);
         } else {
-            $this->_cache[$pasteid] = array('data' => $paste['data']);
+            self::$_cache[$pasteid] = array('data' => $paste['data']);
             list($createdKey)       = $this->_getVersionedKeys(1);
         }
 
@@ -255,31 +255,31 @@ class Database extends AbstractData
             $paste['meta'] = array();
         }
         $paste                                       = self::upgradePreV1Format($paste);
-        $this->_cache[$pasteid]['meta']              = $paste['meta'];
-        $this->_cache[$pasteid]['meta'][$createdKey] = (int) $paste['postdate'];
+        self::$_cache[$pasteid]['meta']              = $paste['meta'];
+        self::$_cache[$pasteid]['meta'][$createdKey] = (int) $paste['postdate'];
         $expire_date                                 = (int) $paste['expiredate'];
         if ($expire_date > 0) {
-            $this->_cache[$pasteid]['meta']['expire_date'] = $expire_date;
+            self::$_cache[$pasteid]['meta']['expire_date'] = $expire_date;
         }
         if ($isVersion2) {
-            return $this->_cache[$pasteid];
+            return self::$_cache[$pasteid];
         }
 
         // support v1 attachments
         if (array_key_exists('attachment', $paste) && !empty($paste['attachment'])) {
-            $this->_cache[$pasteid]['attachment'] = $paste['attachment'];
+            self::$_cache[$pasteid]['attachment'] = $paste['attachment'];
             if (array_key_exists('attachmentname', $paste) && !empty($paste['attachmentname'])) {
-                $this->_cache[$pasteid]['attachmentname'] = $paste['attachmentname'];
+                self::$_cache[$pasteid]['attachmentname'] = $paste['attachmentname'];
             }
         }
         if ($paste['opendiscussion']) {
-            $this->_cache[$pasteid]['meta']['opendiscussion'] = true;
+            self::$_cache[$pasteid]['meta']['opendiscussion'] = true;
         }
         if ($paste['burnafterreading']) {
-            $this->_cache[$pasteid]['meta']['burnafterreading'] = true;
+            self::$_cache[$pasteid]['meta']['burnafterreading'] = true;
         }
 
-        return $this->_cache[$pasteid];
+        return self::$_cache[$pasteid];
     }
 
     /**
@@ -299,9 +299,9 @@ class Database extends AbstractData
             '" WHERE "pasteid" = ?', array($pasteid)
         );
         if (
-            array_key_exists($pasteid, $this->_cache)
+            array_key_exists($pasteid, self::$_cache)
         ) {
-            unset($this->_cache[$pasteid]);
+            unset(self::$_cache[$pasteid]);
         }
     }
 
@@ -315,11 +315,11 @@ class Database extends AbstractData
     public function exists($pasteid)
     {
         if (
-            !array_key_exists($pasteid, $this->_cache)
+            !array_key_exists($pasteid, self::$_cache)
         ) {
-            $this->_cache[$pasteid] = $this->read($pasteid);
+            self::$_cache[$pasteid] = $this->read($pasteid);
         }
-        return (bool) $this->_cache[$pasteid];
+        return (bool) self::$_cache[$pasteid];
     }
 
     /**
@@ -870,10 +870,11 @@ class Database extends AbstractData
      * From: https://stackoverflow.com/questions/36200534/pdo-oci-into-a-clob-field
      *
      * @access public
+     * @static
      * @param  int|string|resource $value
      * @return int|string
      */
-    public function _sanitizeClob($value)
+    public static function _sanitizeClob($value)
     {
         if (is_resource($value)) {
             $value = stream_get_contents($value);
```

I get the GCS tests to pass when making the $client static. I don't know the GCS library enough to understand if that would cause a problem for a GCS to GCS migration or not. The bucket would still get reset in each of the instances - would that work with the migration script? It seems to undermine switching away from a singleton...
```diff
diff --git a/lib/Data/GoogleCloudStorage.php b/lib/Data/GoogleCloudStorage.php
index dd2aff3..55a0013 100644
--- a/lib/Data/GoogleCloudStorage.php
+++ b/lib/Data/GoogleCloudStorage.php
@@ -16,7 +16,7 @@ class GoogleCloudStorage extends AbstractData
      * @access private
      * @var    StorageClient
      */
-    private $_client = null;
+    private static $_client = null;
 
     /**
      * GCS bucket
@@ -64,10 +64,12 @@ class GoogleCloudStorage extends AbstractData
             $this->_uniformacl = $options['uniformacl'];
         }
 
-        $this->_client = class_exists('StorageClientStub', false) ?
-            new \StorageClientStub(array()) :
-            new StorageClient(array('suppressKeyFileNotice' => true));
-        $this->_bucket = $this->_client->bucket($bucket);
+        if (empty(self::$_client)) {
+            self::$_client = class_exists('StorageClientStub', false) ?
+                new \StorageClientStub(array()) :
+                new StorageClient(array('suppressKeyFileNotice' => true));
+        }
+        $this->_bucket = self::$_client->bucket($bucket);
     }
 
     /**
```

**Comment by felixjogris** [comment]
> For the Database the solution is to keep the cache static:

If we keep the cache at all, database to database migrations will always fail. Database->create() checks whether the paste id already exists in the cache. read() places all paste ids in the cache.
So, even if we write to the new database after all pastes have been read to memory from the source database, migration will fail.
After all, is a local cache just for pastes in the database backend really needed? It doesn't cover comments. And Filesystem nor GCS backends implement a cache.
Further more, one can trash the cache be querying for non-existing paste ids: https://github.com/PrivateBin/PrivateBin/blob/master/lib/Data/Database.php#L240 


> I get the GCS tests to pass when making the $client static. I don't know the GCS library enough to understand if that would cause a problem for a GCS to GCS migration or not. The bucket would still get reset in each of the instances - would that work with the migration script? It seems to undermine switching away from a singleton...

Took me a while, but if we change _buckets in https://github.com/PrivateBin/PrivateBin/blob/master/tst/Bootstrap.php#L37 to static instead, tests pass ok. GCS backend "talks" to a StorageClientStub for lack of a real storage connection. Each instance of GoogleCloudStorage creates its own StorageClientStub while testing (https://github.com/PrivateBin/PrivateBin/blob/master/lib/Data/GoogleCloudStorage.php#L80). But these StorageClientStub don't share any data.

**Review by elrido — APPROVED** [review]
Thank you, removing the cache also reduces complexity and it would not have gotten a lot of use under normal operation, where the objects don't outlive the requests and each request only does one operation at a time.

**Pull request merged**